src/SFA.DAS.EmployerAccounts.Jobs/RunOnceJobs/SeedAccountSignedAgreementsJob.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using AutoMapper;
+using AutoMapper.QueryableExtensions;
+using Microsoft.Azure.WebJobs;
+using Microsoft.Extensions.Logging;
+using SFA.DAS.EmployerAccounts.Data;
+using SFA.DAS.EmployerAccounts.Dtos;
+using SFA.DAS.EmployerAccounts.ReadStore.Data;
+using SFA.DAS.EmployerAccounts.ReadStore.Models;
+
+namespace SFA.DAS.EmployerAccounts.Jobs.RunOnceJobs
+{
+    public class SeedAccountSignedAgreementsJob
+    {
+        private readonly IRunOnceJobsService _runOnceJobsService;
+        private readonly IAccountSignedAgreementsRepository _accountSignedAgreementsRepository;
+        private readonly Lazy<EmployerAccountsDbContext> _db;
+        private readonly ILogger _logger;
+        private readonly string _jobName;
+        private readonly IConfigurationProvider _configurationProvider;
+
+        public SeedAccountSignedAgreementsJob(IRunOnceJobsService runOnceJobsService, IAccountSignedAgreementsRepository accountSignedAgreementsRepository, Lazy<EmployerAccountsDbContext> db, ILogger logger, IConfigurationProvider configurationProvider)
+        {
+            _runOnceJobsService = runOnceJobsService;
+            _accountSignedAgreementsRepository = accountSignedAgreementsRepository;
+            _db = db;
+            _logger = logger;
+            _configurationProvider = configurationProvider;
+            _jobName = typeof(SeedAccountSignedAgreementsJob).Name;
+        }
+
+        [NoAutomaticTrigger]
+        public Task Run()
+        {
+            return _runOnceJobsService.RunOnce(_jobName, MigrateAgreements);
+        }
+
+        public async Task MigrateAgreements()
+        {
+            var agreements = _db.Value.Agreements
+                .ProjectTo<AgreementDto>(_configurationProvider)
+                .Where(x => x.SignedDate != null);
+
+            _logger.LogInformation("Migrating signed agreements into the read store");
+
+            foreach (var agreement in agreements)
+            {
+                var agreementExists = await CosmosDb.QueryableExtensions.AnyAsync(
+                    _accountSignedAgreementsRepository.CreateQuery(),
+                    x => x.AccountId == agreement.AccountId && x.AgreementType == agreement.Template.AgreementType.ToString() && x.AgreementVersion == agreement.Template.VersionNumber);
+
+                if (!agreementExists)
+                {
+                    var document = new AccountSignedAgreement(agreement.AccountId, agreement.Template.VersionNumber, agreement.Template.AgreementType.ToString(), Guid.NewGuid());
+                    await _accountSignedAgreementsRepository.Add(document, null, CancellationToken.None);
+                }
+            }
+
+            _logger.LogInformation("Finished migrating signed agreements into the read store");
+        }
```

Comment:
So this is something a little off the cuff to change the approach. It would need testing but illustrates the idea of rather than retrieving then looping over 1 by 1, we retrieve the data sets to do a local join and then can create the ones that are needed. 

Something to consider if we think the data set will be too large as we go through the environments.

```
public async Task MigrateAgreements()
        {
            var agreements = _db.Value.Agreements
                .ProjectTo<AgreementDto>(_configurationProvider)
                .Where(x => x.SignedDate != null).ToList();

            _logger.LogInformation("Migrating signed agreements into the read store");

            var financeAgreement = await CosmosDb.QueryableExtensions.ToListAsync(_accountSignedAgreementsRepository.CreateQuery());

            var existingAgreements = from a in agreements
                                     join f in financeAgreement
                                     on a.AccountId equals f.AccountId
                                     where a.Template.AgreementType.ToString() == f.AgreementType && a.Template.VersionNumber == f.AgreementVersion
                                     select a.Id;

            var agreementsToCreate = agreements.Where(s => !existingAgreements.Contains(s.Id));
            foreach (var agreement in agreementsToCreate)
            {
                var document = new AccountSignedAgreement(agreement.AccountId, agreement.Template.VersionNumber, agreement.Template.AgreementType.ToString(), Guid.NewGuid());
                await _accountSignedAgreementsRepository.Add(document, null, CancellationToken.None);
            }

            _logger.LogInformation("Finished migrating signed agreements into the read store");
        }
```